### DISCUSSION TITLE
import external texture created by java

### DISCUSSION THREAD

**Discussion by FinleyPan** [google/filament#5664] [General]
Hey, I want to know whether feasible importing external textures into filament which created by calling android GLES API:
```java
    //createExternalTexture.java
    public static int createTexture(int width, int height, int pixelFormat) {
        int[] texture = new int[1];

        GLES30.glGenTextures(1, texture, 0);
        GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, texture[0]);
        GLES30.glTexImage2D(GLES30.GL_TEXTURE_2D, 0, pixelFormat, width, height,
                0, pixelFormat, GLES30.GL_UNSIGNED_BYTE, null);

        GLES30.glTexParameterf(GLES30.GL_TEXTURE_2D, GL10.GL_TEXTURE_MIN_FILTER,GL10.GL_LINEAR);
        GLES30.glTexParameterf(GLES30.GL_TEXTURE_2D, GL10.GL_TEXTURE_MAG_FILTER, GL10.GL_LINEAR);
        GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GL10.GL_TEXTURE_WRAP_S, GL10.GL_CLAMP_TO_EDGE);
        GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GL10.GL_TEXTURE_WRAP_T, GL10.GL_CLAMP_TO_EDGE);
        GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, 0);
        return texture[0];
    }
...
    int importTexId = createTexture(width, height, GLES30.GL_RGBA)
```

```cpp
//importExternalTexture.cpp
    filamentTex = filament::Texture::Builder()
            .width((uint32_t)width)
            .height((uint32_t)height)
            .levels(1)
            .sampler(filament::Texture::Sampler::SAMPLER_2D)
            .format(filament::Texture::InternalFormat::RGBA8)
            .usage(filament::Texture::Usage::SAMPLEABLE | filament::Texture::Usage::COLOR_ATTACHMENT)
            .import(importTexId)
            .build(*mEngine);
```

**Comment by romainguy** [comment]
You can import a texture but there's another (I would suggest better) way which is to use a SurfaceTexture.

**Reply by FinleyPan** [reply]
I am really a novice for surface and surfaceview, here is the codes I create the surface and surfacetexture:
```java
        mTexId = createTexture(width, height, GLES20.GL_RGBA);
        mTexture = new SurfaceTexture(mTexId);
        mTexture.setDefaultBufferSize(width, height);
        mSurface = new Surface(mTexture);
```
so is the better way importing the `mTexId` into filament?

**Reply by FinleyPan** [reply]
I use `mSurface` as filament's swapchain

**Reply by romainguy** [reply]
Well what do you want to do with that texture? If you want Filament to render into it, what you're doing is the right approach.

**Reply by FinleyPan** [reply]
Sorry, I want sample the texture

**Reply by FinleyPan** [reply]
can you show me some lines of psuedo codes for importing a surfacetexture?

**Reply by romainguy** [reply]
It's achieved by creating an external texture on the filament end, as well as a Stream. You want tbe Stream to be of type NATIVE and you give it the SurfaceTexture as the stream source. See the javadoc here: https://github.com/google/filament/blob/main/android/filament-android/src/main/java/com/google/android/filament/Stream.java

You can also refer to this example: https://github.com/google/filament/tree/main/android/samples/sample-stream-test

**Reply by FinleyPan** [reply]
cool, thanks!

**Comment by dandingol03** [comment]
> Hey, I want to know whether feasible importing external textures into filament which created by calling android GLES API:
> 
> ```java
>     //createExternalTexture.java
>     public static int createTexture(int width, int height, int pixelFormat) {
>         int[] texture = new int[1];
> 
>         GLES30.glGenTextures(1, texture, 0);
>         GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, texture[0]);
>         GLES30.glTexImage2D(GLES30.GL_TEXTURE_2D, 0, pixelFormat, width, height,
>                 0, pixelFormat, GLES30.GL_UNSIGNED_BYTE, null);
> 
>         GLES30.glTexParameterf(GLES30.GL_TEXTURE_2D, GL10.GL_TEXTURE_MIN_FILTER,GL10.GL_LINEAR);
>         GLES30.glTexParameterf(GLES30.GL_TEXTURE_2D, GL10.GL_TEXTURE_MAG_FILTER, GL10.GL_LINEAR);
>         GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GL10.GL_TEXTURE_WRAP_S, GL10.GL_CLAMP_TO_EDGE);
>         GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GL10.GL_TEXTURE_WRAP_T, GL10.GL_CLAMP_TO_EDGE);
>         GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, 0);
>         return texture[0];
>     }
> ...
>     int importTexId = createTexture(width, height, GLES30.GL_RGBA)
> ```
> 
> ```c++
> //importExternalTexture.cpp
>     filamentTex = filament::Texture::Builder()
>             .width((uint32_t)width)
>             .height((uint32_t)height)
>             .levels(1)
>             .sampler(filament::Texture::Sampler::SAMPLER_2D)
>             .format(filament::Texture::InternalFormat::RGBA8)
>             .usage(filament::Texture::Usage::SAMPLEABLE | filament::Texture::Usage::COLOR_ATTACHMENT)
>             .import(importTexId)
>             .build(*mEngine);
> ```

I can't use importtexture to import an opengles 2d texture. Is it convinient for you to share the whole code? Thanks in advance
